File: lib-database-entities/src/main/java/com/ustadmobile/lib/db/entities/LanguageVariant.java
package com.ustadmobile.lib.db.entities;

import com.ustadmobile.lib.database.annotation.UmEntity;
import com.ustadmobile.lib.database.annotation.UmPrimaryKey;
import com.ustadmobile.lib.database.annotation.UmSyncLocalChangeSeqNum;
import com.ustadmobile.lib.database.annotation.UmSyncMasterChangeSeqNum;

import static com.ustadmobile.lib.db.entities.LanguageVariant.TABLE_ID;


@UmEntity(tableId = TABLE_ID)
public class LanguageVariant {

    public static final int TABLE_ID = 10;

    @UmPrimaryKey(autoGenerateSyncable = true)
    private long langVariantUid;

    private long langUid;

    private String countryCode;

    private String name;

    @UmSyncLocalChangeSeqNum
    private long langVariantLocalChangeSeqNum;

    @UmSyncMasterChangeSeqNum
    private long langVariantMasterChangeSeqNum;


    public long getLangVariantUid() {
        return langVariantUid;
    }

    public void setLangVariantUid(long langVariantUid) {
        this.langVariantUid = langVariantUid;
    }

    public long getLangUid() {
        return langUid;
    }

    public void setLangUid(long langUid) {
        this.langUid = langUid;
    }

    public String getCountryCode() {
        return countryCode;
    }

    public void setCountryCode(String countryCode) {
        this.countryCode = countryCode;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public long getLangVariantLocalChangeSeqNum() {
        return langVariantLocalChangeSeqNum;
    }

    public void setLangVariantLocalChangeSeqNum(long langVariantLocalChangeSeqNum) {
        this.langVariantLocalChangeSeqNum = langVariantLocalChangeSeqNum;
    }

    public long getLangVariantMasterChangeSeqNum() {
        return langVariantMasterChangeSeqNum;
    }

    public void setLangVariantMasterChangeSeqNum(long langVariantMasterChangeSeqNum) {
        this.langVariantMasterChangeSeqNum = langVariantMasterChangeSeqNum;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        LanguageVariant that = (LanguageVariant) o;

        if (langVariantUid != that.langVariantUid) return false;
        if (langUid != that.langUid) return false;
        if (countryCode != null ? !countryCode.equals(that.countryCode) : that.countryCode != null)
            return false;
        return name != null ? name.equals(that.name) : that.name == null;
    }

    @Override
    public int hashCode() {
        int result = (int) (langVariantUid ^ (langVariantUid >>> 32));
        result = 31 * result + (int) (langUid ^ (langUid >>> 32));
        result = 31 * result + (countryCode != null ? countryCode.hashCode() : 0);
        result = 31 * result + (name != null ? name.hashCode() : 0);
        return result;
    }
}

File: lib-database-entities/src/main/java/com/ustadmobile/lib/db/entities/LanguageVariant.kt
package com.ustadmobile.lib.db.entities

import com.ustadmobile.lib.database.annotation.UmEntity
import com.ustadmobile.lib.database.annotation.UmPrimaryKey
import com.ustadmobile.lib.database.annotation.UmSyncLastChangedBy
import com.ustadmobile.lib.database.annotation.UmSyncLocalChangeSeqNum
import com.ustadmobile.lib.database.annotation.UmSyncMasterChangeSeqNum

import com.ustadmobile.lib.db.entities.LanguageVariant.Companion.TABLE_ID


@UmEntity(tableId = TABLE_ID)
class LanguageVariant {

    @UmPrimaryKey(autoGenerateSyncable = true)
    var langVariantUid: Long = 0

    var langUid: Long = 0

    var countryCode: String? = null

    var name: String? = null

    @UmSyncLocalChangeSeqNum
    var langVariantLocalChangeSeqNum: Long = 0

    @UmSyncMasterChangeSeqNum
    var langVariantMasterChangeSeqNum: Long = 0

    @UmSyncLastChangedBy
    var langVariantLastChangedBy: Int = 0

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val that = other as LanguageVariant?

        if (langVariantUid != that!!.langVariantUid) return false
        if (langUid != that.langUid) return false
        if (if (countryCode != null) countryCode != that.countryCode else that.countryCode != null)
            return false
        return if (name != null) name == that.name else that.name == null
    }

    override fun hashCode(): Int {
        var result = (langVariantUid xor langVariantUid.ushr(32)).toInt()
        result = 31 * result + (langUid xor langUid.ushr(32)).toInt()
        result = 31 * result + if (countryCode != null) countryCode!!.hashCode() else 0
        result = 31 * result + if (name != null) name!!.hashCode() else 0
        return result
    }

    companion object {

        const val TABLE_ID = 10
    }
}
